### ISSUE TITLE
Why is the histogram shifted twice?

### ISSUE THREAD

**Issue by mertkaraoglu** [bluedream1121/RELF#3]
Thank you for the making the source code available. In the paper, it is mentioned that only the descriptor 2 (of warped image) is shifted with the known rotation value to align with the descriptor 1 and on this is the alignment loss computed. However, if I am not missing a point, in the code it is applied twice:

1. at [train.py line:47](https://github.com/bluedream1121/RELF/blob/main/train.py#L47) on descriptor 2 towards left;
2. at [train.py line:50](https://github.com/bluedream1121/RELF/blob/main/train.py#L50) on descriptor 1 towards right.

Can you explain why is this the case?

**Comment by ZhHaoH** [comment]
> Thank you for the making the source code available. In the paper, it is mentioned that only the descriptor 2 (of warped image) is shifted with the known rotation value to align with the descriptor 1 and on this is the alignment loss computed. However, if I am not missing a point, in the code it is applied twice:
> 
> 1. at [train.py line:47](https://github.com/bluedream1121/RELF/blob/main/train.py#L47) on descriptor 2 towards left;
> 2. at [train.py line:50](https://github.com/bluedream1121/RELF/blob/main/train.py#L50) on descriptor 1 towards right.
> 
> Can you explain why is this the case?

I also want to know the answer to this question. By the way, may I ask if the value of ori_loss changes when training this network? During my training, the value of ori_loss hardly changes; it consistently decreases the value of desc_loss.

**Comment by mertkaraoglu** [comment]
> > Thank you for the making the source code available. In the paper, it is mentioned that only the descriptor 2 (of warped image) is shifted with the known rotation value to align with the descriptor 1 and on this is the alignment loss computed. However, if I am not missing a point, in the code it is applied twice:
> > 
> > 1. at [train.py line:47](https://github.com/bluedream1121/RELF/blob/main/train.py#L47) on descriptor 2 towards left;
> > 2. at [train.py line:50](https://github.com/bluedream1121/RELF/blob/main/train.py#L50) on descriptor 1 towards right.
> > 
> > Can you explain why is this the case?
> 
> I also want to know the answer to this question. By the way, may I ask if the value of ori_loss changes when training this network? During my training, the value of ori_loss hardly changes; it consistently decreases the value of desc_loss.

Yes, that's also the case here.

**Comment by StefanBst** [comment]
> > > Thank you for the making the source code available. In the paper, it is mentioned that only the descriptor 2 (of warped image) is shifted with the known rotation value to align with the descriptor 1 and on this is the alignment loss computed. However, if I am not missing a point, in the code it is applied twice:
> > > 
> > > 1. at [train.py line:47](https://github.com/bluedream1121/RELF/blob/main/train.py#L47) on descriptor 2 towards left;
> > > 2. at [train.py line:50](https://github.com/bluedream1121/RELF/blob/main/train.py#L50) on descriptor 1 towards right.
> > > 
> > > Can you explain why is this the case?
> > 
> > 
> > I also want to know the answer to this question. By the way, may I ask if the value of ori_loss changes when training this network? During my training, the value of ori_loss hardly changes; it consistently decreases the value of desc_loss.
> 
> Yes, that's also the case here.

Hi! 
probably a bit late but could you find out why they shift desc1 in the alignment loss again?
I changed the loss to: 
     #desc1_shift = torch.roll(desc1[b, :, 0, :], shifts=int(shift), dims=1)
     desc1_shift = desc1[b,:,0,:]
and it seems like it makes almost no difference.

**Comment by mertkaraoglu** [comment]
> > > > Thank you for the making the source code available. In the paper, it is mentioned that only the descriptor 2 (of warped image) is shifted with the known rotation value to align with the descriptor 1 and on this is the alignment loss computed. However, if I am not missing a point, in the code it is applied twice:
> > > > 
> > > > 1. at [train.py line:47](https://github.com/bluedream1121/RELF/blob/main/train.py#L47) on descriptor 2 towards left;
> > > > 2. at [train.py line:50](https://github.com/bluedream1121/RELF/blob/main/train.py#L50) on descriptor 1 towards right.
> > > > 
> > > > Can you explain why is this the case?
> > > 
> > > 
> > > I also want to know the answer to this question. By the way, may I ask if the value of ori_loss changes when training this network? During my training, the value of ori_loss hardly changes; it consistently decreases the value of desc_loss.
> > 
> > 
> > Yes, that's also the case here.
> 
> Hi! probably a bit late but could you find out why they shift desc1 in the alignment loss again? I changed the loss to: #desc1_shift = torch.roll(desc1[b, :, 0, :], shifts=int(shift), dims=1) desc1_shift = desc1[b,:,0,:] and it seems like it makes almost no difference.

No, not yet. I also did not have the chance to train the model with a fix similar to what you have suggested. Can you perhaps let me know if there is any changes happen to the rotational robustness experiment displayed in Figure 5? Especially its performance for in-between rotations of 22.5 degree (for cyclic group of 16) increments?
